### PR TITLE
Implementation of nextBuildNumber support in the DSL.

### PR DESCRIPTION
**Pull request by akomakom**
Implements proposed changes in https://issues.jenkins-ci.org/browse/JENKINS-30127 to support nextBuildNumber in the DSL:

    freeStyleJob(String name) {
      name(String name)
      nextBuildNumber(int nextBuildNumber)  // This is new
      // etc...
    }

### CODE REVIEW

**Comment by jwhitcraft** [comment]
@daspilker any chance of this getting in the next release?

**Comment by KostyaSha** [comment]
Please create a test case that will show that build number wouldn't be changed for already existed number and that it really sets build number.

File: job-dsl-plugin/src/main/groovy/javaposse/jobdsl/plugin/JenkinsJobManagement.java [hunk 1]
```diff
@@ -381,6 +381,29 @@ public boolean apply(Job topLevelItem) {
     }
 
     @Override
+    public void updateNextBuildNumber(final String jobName, int nextBuildNumber) throws IOException {
+        if (nextBuildNumber <= 0) {
+            return;
+        }
+        LOGGER.fine(getClass().getSimpleName() + ": looking for job  " + jobName + " to update next build number to " + nextBuildNumber);
```
Comment (KostyaSha) [inline job-dsl-plugin/src/main/groovy/javaposse/jobdsl/plugin/JenkinsJobManagement.java]: @daspilker is some advanced logger available in plugin?

**Comment by akomakom** [comment]
@KostyaSha, I added basic tests for set and not negative, although I don't feel that they are of any great benefit.  Unlike most other options in the DSL, nextBuildNumber does not modify the XML and thus cannot be properly tested using MockJobManagement or its descendants.   The value is applied after the job is created in **JenkinsJobManagement**, and it is a jenkins core method and not a job dsl method that enforces **only set if greater or equal** rule.   I can duplicate that rule in MockJobManagement but what would be the point?

**Comment by KostyaSha** [comment]
Why not just create a test with JenkinsRule? Job with very simple DSL and 2 or 3 build schedules with build number verifications. 
I fear that issue may appear when i apply this change to existed job and it will decrease number, that will cause conflicts.

**Comment by akomakom** [comment]
I'll look into JenkinsRule.   
As to your concern, that is impossible based on the jenkins core code that I'm calling. 
hudson.model.Job:

```java
    /**
     * Programatically updates the next build number.
     * 
     * <p>
     * Much of Hudson assumes that the build number is unique and monotonic, so
     * this method can only accept a new value that's bigger than
     * {@link #getLastBuild()} returns. Otherwise it'll be no-op.
     * 
     * @since 1.199 (before that, this method was package private.)
     */
    public synchronized void updateNextBuildNumber(int next) throws IOException {
        RunT lb = getLastBuild();
        if (lb!=null ?  next>lb.getNumber() : next>0) {
            this.nextBuildNumber = next;
            saveNextBuildNumber();
        }
    }
```

**Comment by KostyaSha** [comment]
That's good, but some test case imho will be useful.

**Comment by akomakom** [comment]
@KostyaSha, can you suggest an example of using JenkinsRule to accomplish this?  Do I need to start a fake Jenkins and install the DSL plugin?  I don't see how using any of the available MockJobManagement implementations could affect state represented in an instance of JenkinsRule.

Alternatively, feel free to contribute such a test to the pull request.

**Comment by akomakom** [comment]
@KostyaSha I added the test you requested

**Comment by jwhitcraft** [comment]
Can this be merged now?

**Comment by daspilker** [comment]
What's the use case for setting the build number with the Job DSL plugin? Isn't the [Description Setter Plugin](https://wiki.jenkins-ci.org/display/JENKINS/Description+Setter+Plugin) or the [Build Name Setter Plugin](https://wiki.jenkins-ci.org/display/JENKINS/Build+Name+Setter+Plugin) the better option to get a nice build number?

It's a transient, one-time setting which gets meaningless and confusing after the first seed job run. I would prefer to keep build-level settings out of the core DSL.

@akomakom what happened to your [approach](https://groups.google.com/d/msg/job-dsl-plugin/VybUIlHbLac/PvzzU7nrAwAJ) to integrate this into the Next Build Number Plugin?

**Comment by akomakom** [comment]
@daspilker :  

First, the use case:
* Builds produce artifacts tied to build numbers, therefore a job should never run with an older build number.  That should be typical.
* I want to be able to tear down the job (a whole pipeline of 30 jobs, in my case) and re-create from scratch, possibly on a different master.   When I do that, I should be able to set the initial build numbers for obvious reasons.  Yes, I can write custom code to curl the Next Build Number action (and fight with authentication, etc), but I'm using the DSL as a one-stop solution for my needs which I think is reasonable.  It is also quite a bit easier than editing the buildNumber file and restarting the master.

I agree that the value is meaningless after the first run, if you are re-running the seed job (unless you are rerunning it to change the number), but the safety it provides with regard to the above outweighs the oddity in my mind.

Unless I'm misunderstanding something, the two plugins you cited do not provide this functionality.

As to your question regarding integrating this into the Next Build Number Plugin, I closed my pull request after a conversation with its owner.   It was too hard to do via the extension points (I had to do it as a trigger which was a horrific hack), and in any case it would still wind up being a DSL feature, which it would seem is what you are opposing.

**Comment by daspilker** [comment]
I don't like this as a *core* DSL feature, but plugins could add such features.

I think doing this as an extension point is hard, because there is no need to generate any config XML. But a DslExtensionMethod is expected to return something to contribute to the XML.

If I change the extension point, the implementation could be as simple as this (might not compile, just to get the idea):

```java
@Extension(optional = true)
public class NextBuildNumberDslExtension extends ContextExtensionPoint {
    @DslExtensionMethod(context = PropertiesContext.class)
    public Object nextBuildNumber(int number, DslEnvironment dslEnvironment) {
        dslEnvironment.put('nextBuildNumber', number);
        return NO_VALUE; // special marker to indicate that no XML should be generated
    }

    @Override
    public void notifyItemCreated(Item item,
                                  DslEnvironment dslEnvironment) {
        notifyItemUpdated(item, dslEnvironment);
    }

    @Override
    public void notifyItemUpdated(Item item,
                                  DslEnvironment dslEnvironment) {
        if (item instance of Job && dslEnvironment.containsKey('nextBuildNumber')) {
          ((Job) item).updateNextBuildNumber(dslEnvironment.get('nextBuildNumber'));
        }
    }
}
```

**Comment by jwhitcraft** [comment]
@daspilker is there any docs on how to add plugins?

**Comment by daspilker** [comment]
Replaced by https://github.com/jenkinsci/next-build-number-plugin/pull/4.

**Pull request closed without merging**